Refactor Buffer Placement MILP Constraints for Modular Extension

QinYuan2000 opened the pull request:
### Summary
This PR is part of the work towards issue [#228](https://github.com/EPFL-LAP/dynamatic/issues/228). It focuses on refactoring and modularizing the MILP constraints to support future MILP variants.

### Details
1. **Refactored Constraints**
   - The constraints function in `BufferPlacementMILP` has been split into smaller, focused functions.
   - This makes the code easier to extend, as different MILPs can now selectively use specific parts of the constraints.

2. **Channel Delay Constraint Variants**
   - Some MILPs do not model buffer delay, while others do.
   - Two versions of the channel delay constraint are now available:
     - One that does **not** include buffer delay.
     - One that **does** include buffer delay.
   - Each MILP can choose which version to apply.
   - This serves as the first example of future variants.

3. **Naming Consistency**
   - Standardized input variables of type `SignalType` to use `signal` as the name.
   - In structs, kept full names like `channelVars` and `cfdfcVars`.
   - Inside functions, used short names like `chVars` and `cfVars` for local references.

### Outlook
Future PRs will add MILP variants that build on these refactored constraints. This structure allows easier extension and cleaner integration of new features.

Jiahui17 commented:
@QinYuan2000 It is easier to review if you only include the splitting part

QinYuan2000 commented:
> @QinYuan2000 It is easier to review if you only include the splitting part

Updated, thanks!

Jiahui17 commented on lib/Transforms/BufferPlacement/BufferPlacementMILP.cpp:
```diff
@@ -373,7 +391,9 @@ void BufferPlacementMILP::addUnitElasticityConstraints(Operation *unit,
 }
 
-void BufferPlacementMILP::addChannelThroughputConstraints(CFDFC &cfdfc) {
+// Generate token occupancy distribution which represents a steady state
```
Be precise:

// Constraints that ensure the final throughput is calculated for a reachable steady state.

So maybe I would call it "addSteadyStateReachabilityConstraint"

QinYuan2000 replied:
Sure, thank you.

Jiahui17 commented on lib/Transforms/BufferPlacement/BufferPlacementMILP.cpp:
```diff
@@ -415,6 +427,44 @@ void BufferPlacementMILP::addChannelThroughputConstraints(CFDFC &cfdfc) {
+  }
+}
+
+// Only channel throughput constraints. They are linear.
+void BufferPlacementMILP::addBasicChannelThroughputConstraints(CFDFC &cfdfc) {
```
I don't like the word "basic" or "linear": it doesn't tell us what it actually is.

I think we should emphasize that we write (linearize) the constraint in this way because we know that the channel latency is a binary value.

So, maybe we can use: `addChannelThroughputConstraintForBinaryLatencyChannel`

QinYuan2000 replied:
Sure, updates have been made.

QinYuan2000 commented on lib/Transforms/BufferPlacement/BufferPlacementMILP.cpp:
```diff
@@ -310,31 +310,32 @@ void BufferPlacementMILP::addUnitPathConstraints(Operation *unit,
-    Value channel, ArrayRef<BufferingGroup> bufGroups) {
+// This function shows the relationship among buffer presence, signal
+// buffer presence, and the number of slots on the channel.
+void BufferPlacementMILP::addBufferPresenceConstraints(Value channel) {
```
Sure, I’ve added this description to the header file, alongside the other function descriptions.

QinYuan2000 commented on lib/Transforms/BufferPlacement/BufferPlacementMILP.cpp:
```diff
@@ -358,7 +359,24 @@ void BufferPlacementMILP::addChannelElasticityConstraints(
 
-void BufferPlacementMILP::addUnitElasticityConstraints(Operation *unit,
+// The follwing two constraints are used to break combinational cycles in the CFG
+void BufferPlacementMILP::addBreakingCycleChannelConstraints(Value channel) {
```
I’m not sure whether they can be removed. I’ve described their effect in the header file, and also revised the function name.

Jiahui17 replied:
I am kind of confident that they can be removed :) Should we try to see if the integration tests work fine?

QinYuan2000 replied:
After a few tests, the zero-delay combinational cycle exists and clearly affects buffer placement. For example, in the `fir` case, there is a cycle consisting only of `cmerge` and `cond_br`, both of which have zero data-to-data delay. They are circled in black in the figure below:

<p float="left">
  <img src="https://github.com/user-attachments/assets/55c9638c-68e3-4a1a-8ac6-a4cde5fc91f4" width="400"/>
</p>

If the extra constraints are omitted, no DV buffer will be inserted into this cycle—only an R buffer will be placed—resulting in a combinational cycle in the data signal path, as shown below:

<p float="left">
  <img src="https://github.com/user-attachments/assets/a3947e3a-c9a1-413b-9cb0-9d19e5b1dc8b" width="400"/>
</p>

This reduces the throughput by half. Even if the throughput remains unaffected in some cases, the presence of a combinational cycle in the data path is logically incorrect and should be avoided.

QinYuan2000 replied:
I suggest adding the following snippet to the timing constraints to enforce a positive unit delay, rather than relying on additional cycle-detection functions:

```c++
if (delay == 0.0) {
  delay = 0.001;
}
```

However, this approach is not particularly elegant. The delay must be small enough to avoid affecting the timing model, but not so small that it causes numerical issues for Gurobi.

QinYuan2000 replied:
In the last two commits, I replaced the delay logic with max expressions and removed two unnecessary functions along with their related variables.

Jiahui17 commented:
The title of the PR can be shortened a bit

> Certain constraints previously used to prevent zero-delay combinational cycles have been removed and replaced by a max function, ensuring that all unit delays remain positive.

Break into two sentences (it is not clear from the text what this "max function" is

QinYuan2000 commented:
> The title of the PR can be shortened a bit
> 
> > Certain constraints previously used to prevent zero-delay combinational cycles have been removed and replaced by a max function, ensuring that all unit delays remain positive.
> 
> Break into two sentences (it is not clear from the text what this "max function" is.
> 
> The changes look very good now!

Updated.

I found another issue. 

Positive-latency units break DATA paths but not READY paths. Previously, the `break-cycle-constraints` broke all combinational cycles regardless of whether latency units were already present. As a result, even when a latency unit was sufficient to break the DATA path, an additional buffer was still inserted—unintentionally breaking the READY path as well, since in FPGA20Buffers, DATA and READY buffers are placed together. 

However, if we now rely solely on timing constraints, READY-path combinational cycles will no longer be detected in FPGA20Buffers, because READY paths are not modeled there.

QinYuan2000 commented:
> > The title of the PR can be shortened a bit
> > > Certain constraints previously used to prevent zero-delay combinational cycles have been removed and replaced by a max function, ensuring that all unit delays remain positive.
> > 
> > 
> > Break into two sentences (it is not clear from the text what this "max function" is.
> > The changes look very good now!
> 
> Updated.
> 
> I found another issue.
> 
> Positive-latency units break DATA paths but not READY paths. Previously, the `break-cycle-constraints` broke all combinational cycles regardless of whether latency units were already present. As a result, even when a latency unit was sufficient to break the DATA path, an additional buffer was still inserted—unintentionally breaking the READY path as well, since in FPGA20Buffers, DATA and READY buffers are placed together.
> 
> However, if we now rely solely on timing constraints, READY-path combinational cycles will no longer be detected in FPGA20Buffers, because READY paths are not modeled there.

This issue is elaborated on and fixed in #427. This PR keeps the cycle breaking constraints for now, and #427 will remove them after this PR is merged.

Jiahui17 commented:
Let's wait for CI
